File: back/src/main/java/com/back/standard/recommend/util/CosineSimilarityCalcer.java
package com.back.standard.recommend.util;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

import static com.back.standard.recommend.util.Vector.VectorElement;

public class CosineSimilarityCalcer implements SimilarityCalcer {

    private double vectorASqrMagnitude;
    private double vectorBSqrMagnitude;

    private List<VectorElement> sortedVectorA;
    private List<VectorElement> sortedVectorB;

    private static List<VectorElement> getSortedVector(Vector vector) {
        return vector
                .getVectorElementList()
                .stream()
                .sorted(Comparator
                        .comparingLong(
                                VectorElement::label))
                .toList();
    }

    private static double getSqrMagnitude(List<VectorElement> ratings) {

        return ratings
                .stream()
                .mapToDouble(
                vectorElement -> vectorElement.value() * vectorElement.value())
                .sum();
    }

    public void setVectorA(Vector vector) {

        sortedVectorA = getSortedVector(vector);
        vectorASqrMagnitude = getSqrMagnitude(sortedVectorA);
    }

    public void setVectorB(Vector vector) {
        sortedVectorB = getSortedVector(vector);
        vectorBSqrMagnitude = getSqrMagnitude(sortedVectorB);
    }

    public double getSimilarity() {

        int refA = 0;
        int refB = 0;

        double ret = 0;

        while (refA < sortedVectorA.size() && refB < sortedVectorB.size()) {
            VectorElement ratingA = sortedVectorA.get(refA);
            VectorElement ratingB = sortedVectorB.get(refB);

            if (ratingA.label() < ratingB.label()) {
                refA++;
                continue;
            }
            if (ratingA.label() > ratingB.label()) {
                refB++;
                continue;
            }

            ret += ratingA.value() * ratingB.value();

            refA++;
            refB++;
        }

        return ret / Math.sqrt(vectorASqrMagnitude * vectorBSqrMagnitude);
    }


    public List<Similar> getSimilarList(Vector target, Map<Long, Vector> matrix) {

        List<Similar> similarList = new ArrayList<>();

        setVectorA(target);

        matrix.entrySet().stream()
                .filter(set -> {
                    if (set.getValue() == target) return false;
                    return !set.getValue().isEmpty();
                })
                .forEach(set -> {
                    setVectorB(set.getValue());
                    similarList.add(new Similar(set.getKey(), getSimilarity()));
                });

        similarList.sort(Comparator.comparingDouble(a -> -a.score()));

        return similarList;
    }
}

File: back/src/main/kotlin/com/back/standard/recommend/util/CosineSimilarityCalcer.kt
package com.back.standard.recommend.util

class CosineSimilarityCalcer<I, L> : SimilarityCalcer<I, L> {

    private var vectorASqrMagnitude = 0.0
    private var vectorBSqrMagnitude = 0.0

    private var vectorA: Vector<L>? = null
    private var vectorB: Vector<L>? = null

    override fun setVectorA(vector: Vector<L>) {
        vectorA = vector
        vectorASqrMagnitude = getSqrMagnitude(vector.getVectorElementList())
    }

    override fun setVectorB(vector: Vector<L>) {
        vectorB = vector
        vectorBSqrMagnitude = getSqrMagnitude(vector.getVectorElementList())
    }

    override fun getSimilarity(): Double {
        val a = vectorA ?: return 0.0
        val b = vectorB ?: return 0.0

        var ret = 0.0

        a.getVectorElementList().forEach { element ->
            val otherValue = b.getValue(element.label) ?: return@forEach
            ret += element.value * otherValue
        }

        return ret / Math.sqrt(vectorASqrMagnitude * vectorBSqrMagnitude)
    }

    override fun getSimilarList(target: Vector<L>, matrix: Map<I, Vector<L>>): List<SimilarityCalcer.Similar<I>> {
        setVectorA(target)

        return matrix.entries
            .filter { it.value !== target && !it.value.isEmpty() }
            .map { entry ->
                setVectorB(entry.value)
                SimilarityCalcer.Similar(entry.key, getSimilarity())
            }
            .sortedByDescending { it.score }
    }

    companion object {
        private fun <L> getSqrMagnitude(elements: List<Vector.VectorElement<L>>): Double =
            elements.sumOf { it.value * it.value }
    }
}
